### PR TITLE
Query: Relational: Lays groundwork for removing MARS requirement.

### PR DESCRIPTION
**Pull request by anpete**
- Remove remaining DbDataReader dependencies from query processing - Use ValueBuffer everywhere.
- Removed offsetting from ValueBuffer factory factories.
- Enables dynamic ValueBuffer index re-mapping for FromSql non-composed projection out-of-order scenarios.
- Adds workaround for IX-Async #93
- Enables client Select op over non-composed FromSql queries!

dprun: 100%

### CODE REVIEW

File: src/EntityFramework.Relational/Query/ExpressionTreeVisitors/IncludeExpressionTreeVisitor.cs [hunk 1]
```diff
@@ -236,7 +230,7 @@ var innerJoinExpression
                             targetTableExpression,
                             innerJoinExpression);
 
-                    var readerParameter = Expression.Parameter(typeof(DbDataReader), "dataReader");
+                    var valueBufferParameter = Expression.Parameter(typeof(ValueBuffer), "valueBuffer");
```
Comment (DavidObando) [inline src/EntityFramework.Relational/Query/ExpressionTreeVisitors/IncludeExpressionTreeVisitor.cs]: nit: nameof(ValueBuffer)?

Comment (ajcvickers) [inline src/EntityFramework.Relational/Query/ExpressionTreeVisitors/IncludeExpressionTreeVisitor.cs]: I prefer the correct parameter name casing.

**Comment by ajcvickers** [comment]
Can you go into SqlServerDataStoreServices and change UntypedValueBufferFactoryFactory to TypedValueBufferFactoryFactory and then run all the tests. This will test that the typed version of the factory (needed for SQLite, etc.) is still working. It doesn't have full coverage in other ways yet because SQLite tests are still being fixed due to other issues.

**Comment by ajcvickers** [comment]
:shipit:

**Pull request merged**